Some effect cleanups

Paul-Licameli opened the pull request:
Resolves: *(direct link to the issue)*

Some reorganization of effect code, before those that have an "Init()" might also become "PerTrackEffects"
(such as EQ and Noise Reduction), and thus possible to put into the sidebar.

UI-related abstract base classes are moved out of the lower-level libraries that should be concerned only with
calculations.

There is still some state in the Effect objects (even in so-called "stateless" effects) related to the overall context
of tracks and open windows.  This begins to eliminate some of that other state.

QA:  please sanity-check:

Rendering of effect stacks
Destructive application of built-ins and Ladspa effects (don't need many variations of settings)


<!-- Use "x" to fill the checkboxes below like [x] -->

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

vsverchinsky commented on src/effects/Effect.h:
```diff
@@ -180,11 +180,13 @@ class AUDACITY_DLL_API Effect /* not final */
 
-   bool EnableApply(bool enable = true);
+   static
+   bool EnableApply(wxWindow *parent, bool enable = true);
```
As a members those functions make sense here, but not much sense as a statics. I'd suggest to move them out from these class completely

Paul-Licameli replied:
I don't know why it's so bad to make them statics, but as functions related specially to user interface, they might make more sense as static utility members of EffectUIValidator.

vsverchinsky reviewed (commented):
Few suggestions, there is a build error that should be fixed

Paul-Licameli commented on src/effects/Effect.h:
```diff
@@ -304,12 +304,9 @@ class AUDACITY_DLL_API Effect /* not final */
-   wxWindow* GetUIParent() { return mUIParent; }
 
 private:
+   wxWindow       *mUIParent{};
```
I want to eliminiate Effect::mUIParent.  This pull request does not complete that.  It does remove direct uses of it in subclasses, which instead shadow it.  Stateless subclasses store it in the Validator object instead of the effect, which is even better, eliminating some state.

To eliminaite it entirely requires rewriting of Effect::MessageBox and its many uses.  That is to be done later.

Paul-Licameli commented:
I have rebased onto recent master, which had no conflicts.

I added a commit to fix the Windows build, and another to answer the suggestion to move two static member functions out of Effect.

Ryan-Miller-Audacity commented:
Completed sanity checks of mix and rendering of effect stacks, exporting audio, and applying built in/Ladspa effects.
